### PR TITLE
Fill partial fields

### LINKED ISSUE #992: Extend `fill_struct_fields` assist to support partially filled struct literals
We implemented `fill_struct_fields` assist in #964 (PR #989), which was in turn based on #626.

We would like to extend the assist to support partial struct literals. That is, if there is a

```rust
struct S {
    foo: u32,
    bar: i32,
    baz: String
}
```

and a literal `S { bar: 92 }`, the assist should produce `S { foo: (), bar: 92, baz: () }`, preserving the user filled values.

The assist lives in in  [fill_struct_fields.rs](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_assists/src/fill_struct_fields.rs)

### PR DESCRIPTION
**Pull request by c410-f3r**
Fixes #992

### CODE REVIEW

File: crates/ra_assists/src/fill_struct_fields.rs [hunk 1]
```diff
@@ -1,51 +1,99 @@
...
 
-pub(crate) fn fill_struct_fields(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
-    let struct_lit = ctx.node_at_offset::<ast::StructLit>()?;
-    let named_field_list = struct_lit.named_field_list()?;
+struct FillStructFields<'a, 'b: 'a, DB> {
+    borrowed_ctx: &'a mut AssistCtx<'b, DB>,
+    named_field_list: &'a ast::NamedFieldList,
+    struct_fields: Vec<StructField>,
+    struct_lit: &'a ast::StructLit,
+}
```
Comment (matklad) [inline crates/ra_assists/src/fill_struct_fields.rs]: Do we really need this struct though? Can we just turn its methods into functions, which accept a bunch of arguments?

File: crates/ra_hir/src/name.rs [hunk 2]
```diff
@@ -23,7 +23,7 @@ impl fmt::Debug for Name {
 }
 
 impl Name {
-    pub(crate) fn new(text: SmolStr) -> Name {
+    pub fn new(text: SmolStr) -> Name {
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply crates/ra_hir/src/name.rs]: This is unfortunate, `remove_already_included_fields` will be a O(m*n) operation in which every iteration converts a field into a string at run-time. I hope the compiler will optimize it somehow

[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_hir/src/name.rs]: That shouldn't really matter, see https://github.com/rust-analyzer/rust-analyzer/pull/947#discussion_r263446591 :) 

Long-term, we should really be getting structured error from the core, which tells us exactly which fields are missing.

File: crates/ra_assists/src/fill_struct_fields.rs [hunk 3]
```diff
@@ -1,51 +1,99 @@
...
 
 use ra_syntax::ast::{self, AstNode};
 
 use crate::{AssistCtx, Assist, AssistId};
 
-pub(crate) fn fill_struct_fields(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
-    let struct_lit = ctx.node_at_offset::<ast::StructLit>()?;
-    let named_field_list = struct_lit.named_field_list()?;
+struct FillStructFields<'a, 'b: 'a, DB> {
+    borrowed_ctx: &'a mut AssistCtx<'b, DB>,
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply crates/ra_assists/src/fill_struct_fields.rs]: Done

File: crates/ra_assists/src/fill_struct_fields.rs [hunk 4]
```diff
@@ -1,51 +1,99 @@
...
+            },
+        );
+        Some(())
+    }
 
-    // If we already have existing struct fields, don't provide the assist.
-    if named_field_list.fields().count() > 0 {
-        return None;
+    fn check_if_struct_lit_and_def_have_the_same_number_of_fields(&self) -> Option<()> {
+        Some(()).filter(|_| self.named_field_list.fields().count() != self.struct_fields.len())
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply crates/ra_assists/src/fill_struct_fields.rs]: Ok. I mostly did it for an one-line-easier-to-read condition

File: crates/ra_assists/src/fill_struct_fields.rs [hunk 5]
```diff
@@ -1,51 +1,99 @@
...
+        Some(buf)
+    }
+}
 
-        edit.target(struct_lit.syntax().range());
-        edit.set_cursor(struct_lit.syntax().range().start());
-        edit.replace_node_and_indent(named_field_list.syntax(), buf);
-    });
-
+pub(crate) fn fill_struct_fields(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply crates/ra_assists/src/fill_struct_fields.rs]: Done

File: crates/ra_assists/src/fill_struct_fields.rs (hunk 1 above)
Comment (c410-f3r) [reply crates/ra_assists/src/fill_struct_fields.rs]: I did it first time but ended up with a mess of duplicated function signatures

File: crates/ra_hir/src/name.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (c410-f3r) [reply crates/ra_hir/src/name.rs]: Nice!

**Comment by c410-f3r** [comment]
Tweaked code to maintaining the canonical order of the struct fields and improve the overall performance by making less allocations and fewer comparisons.

**Comment by matklad** [comment]
Thanks!

**Comment by yanchith** [comment]
@c410-f3r just wandering, what would the current implementation do when the user filled fields are written in an incorrect order? E.G.

```rust
struct Foo {
    a: u8,
    b: u8,
    c: u8,
}

fn main() {
    let foo = Foo {
        c: 0,
        b: 0,
        <|>
    }
}
```

I don't see any test for it.

**Comment by c410-f3r** [comment]
@yanchith The order is preserved.

To test it, a simple line swap would be needed in `fill_struct_fields_partial`.

https://github.com/rust-analyzer/rust-analyzer/blob/aa0cc0c6096ca8ba80ef225c813fe30310ba6c33/crates/ra_assists/src/fill_struct_fields.rs#L212-L217

**Pull request merged**